File: tests/queue.py
import sys 
import os 

cur_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(cur_dir)
sys.path.insert(0, project_root)

from python.SafeQueue import SafeQueue
import time
import threading 

num_threads = 5
gen_per_thread = 100
log_lock = threading.Lock() 

def log(message):
    with log_lock:
        print(message)

def producer(q, id):
    for i in range(gen_per_thread):
        val = i + (id * gen_per_thread * 10)
        q.push(val)
        
        time.sleep(0.01)
        
    log(f"Producer {id} finished pushing.")

def consumer(q, id):
    while True:
        try:
            item = q.pop()
            log(f"Consumer {id} received {item}.")
        except RuntimeError:
            log(f"Consumer {id} terminated.")
            break

if __name__ == "__main__":
    producers = []
    consumers = []
    queue = SafeQueue()

    for i in range(num_threads):
        t = threading.Thread(target=consumer, args=(queue, i))
        t.start()
        consumers.append(t)

    for i in range(num_threads):
        t = threading.Thread(target=producer, args=(queue, i))
        t.start()
        producers.append(t)

    for t in producers:
        t.join()

    queue.close()

    for t in consumers:
        t.join()
        
    print("All threads finished.")

File: tests/queue.cpp
#include <chrono>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>
#include <vector>
#include "../cpp/SafeQueue.hpp"

std::mutex mtx;

const int num_threads = 5;
const int gen_per_thread = 100;

void log(const std::string& message) {
    std::lock_guard<std::mutex> lock(mtx);
    std::cout << message << std::endl;
}

void producer(SafeQueue<int>& q, int id) {
    for(int i = 0; i < gen_per_thread; i++) {
        q.push(i + id * gen_per_thread * 10);
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    log("Producer " + std::to_string(id) + " finished pushing.");
}

void consumer(SafeQueue<int>& q, int id) {
    while(true) {
        try {
            int item = q.pop();
        } catch (const std::runtime_error& e) {
            log("Consumer " + std::to_string(id) + " terminated.");
            break;
        }
    }
}

int main() {
    std::vector<std::thread> producers;
    std::vector<std::thread> consumers;
    SafeQueue<int> queue;
    for(int i = 0; i < num_threads; i++) {
        consumers.emplace_back(consumer, std::ref(queue), i);
    }

    for(int i = 0; i < num_threads; i++) {
        producers.emplace_back(producer, std::ref(queue), i);
    }

    // produce data
    for(auto &th : producers) {
        if(th.joinable()) th.join();
    }

    // poison pill (indicates producer is done making new instances)
    queue.close();

    // wait for consumers to finish processing
    for(auto &th : consumers) {
        if(th.joinable()) th.join();
    }
}